classes/steps/actions/webservice_action_step.php

```diff
@@ -0,0 +1,162 @@
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_trigger\steps\actions;
+
+defined('MOODLE_INTERNAL') || die();
+
+require_once($CFG->libdir . '/externallib.php');
+
+/**
+ * Webservice action step class.
+ *
+ * @package    tool_trigger
+ * @author     Kevin Pham <kevinpham@catalyst-au.net>
+ * @copyright  Catalyst IT, 2022
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class webservice_action_step extends base_action_step {
+
+    use \tool_trigger\helper\datafield_manager;
+
+    /** @var string $functionname Name of the function to be called */
+    protected $functionname;
+
+    /** @var int $username The user in which this action will be actioned in the context of */
+    protected $username;
+
+    /** @var int $params parameters that will be used in the corresponding web service function call */
+    protected $params;
+
+    /**
+     * The fields supplied by this step.
+     *
+     * @var array
+     */
+    private static $stepfields = [
+        'data',
+        'error',
+        'exception',
+    ];
+
+    protected function init() {
+        $this->functionname = $this->data['functionname'];
+        $this->username = $this->data['username'];
+        $this->params = $this->data['params'];
+    }
+
+    /**
+     * Returns the step name.
+     *
+     * @return string human readable step name.
+     */
+    public static function get_step_name() {
+        return get_string('webserviceactionstepname', 'tool_trigger');
+    }
+
+    /**
+     * Returns the step name.
+     *
+     * @return string human readable step name.
+     */
+    public static function get_step_desc() {
+        return get_string('webserviceactionstepdesc', 'tool_trigger');
+    }
+
+    /**
+     * Runs the configured step.
+     *
+     * @param $trigger
+     * @param $event
+     * @param $stepresults - result of previousstep to include in processing this step.
+     * @return array if execution was succesful and the response from the execution.
+     */
+    public function execute($step, $trigger, $event, $stepresults) {
+        global $CFG, $DB;
+        require_once($CFG->dirroot . '/admin/tool/trigger/guzzle/autoloader.php');
+
+        $this->update_datafields($event, $stepresults);
+
+        // Passing any data from previous steps through by applying template magic.
+        $username = $this->render_datafields($this->username);
+        $functionname = $this->render_datafields($this->functionname);
+        $params = $this->render_datafields($this->params);
+
+        // Assume the role of the provided user given their {username}.
+        $user = $DB->get_record('user', ['username' => $username], '*', MUST_EXIST);
+        $user->ignoresesskey = true; // This bypasses the sesskey check for the external api call.
+        \core\session\manager::set_user($user);
```

Comment:
Just wondering if we should unset the user back to what it was so we don't get any side effects in later steps